### PR TITLE
Add conditional display

### PR DESCRIPTION
**Pull request by mgu**

### CODE REVIEW

**Comment by brunobord** [comment]
You'll need to rebase from master **and** use the ``tox -e swagger-statics`` command in order to make sure the statics for the docs are up-to-date

File: formidable/migrations/0004_formidable_conditions.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('formidable', '0003_item_label_no_size_limit'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='formidable',
+            name='conditions',
+            field=models.TextField(blank=True, null=True),
```
Comment (wo0dyn) [inline formidable/migrations/0004_formidable_conditions.py]: We don't use `JSONField` here as well?

Comment (wo0dyn) [inline formidable/migrations/0004_formidable_conditions.py]: Also, do we need to create a `json_migration`?

Comment (mgu) [inline formidable/migrations/0004_formidable_conditions.py]: we don't need a json_migration, it's a new field

File: demo/tests/test_conditions.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+class ConditionTestCase(TestCase):
+
+    def get_form_class(self, formidable, role):
+        return get_dynamic_form_class(formidable, role)
+
+    def setUp(self):
+        conditions_schema = [
+            {
+                'name': 'My Name',
+                'action': 'display_iff',
```
Comment (brunobord) [inline demo/tests/test_conditions.py]: probably, yes... it'll be less error-prone, especially when we'll have more than this action available.

File: demo/tests/test_conditions.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                'fields_ids': ['foo', 'bar'],
+                'tests': [
+                    {
+                        'field_id': 'checkbox',
+                        'operator': 'eq',
```
Comment (brunobord) [inline demo/tests/test_conditions.py]: same remark for operators, they'll probably need to be categorized.

File: demo/tests/test_end_point.py [hunk 4]
```diff
@@ -652,6 +800,91 @@ def test_create_form_with_presets_invalid_argument(self):
             'Preset (confirmation) argument is using an undefined field (testField2)'  # noqa
         )
 
+    def test_create_form_conditions(self):
```
Comment (brunobord) [inline demo/tests/test_end_point.py]: I would be in favor of a few comments that explain the different steps of this test

File: formidable/forms/conditions.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                                                  test['values']))
+                     for test in condition['tests']]
+            yield klass(
+                condition['fields_ids'],
+                condition['name'],
+                tests
+            )
+
+    def build(self, fields, conditions_schema):
+        # TODO XXX deepcopy ?
```
Comment (brunobord) [inline formidable/forms/conditions.py]: care to explain?

File: formidable/forms/conditions.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+import six
+
+
+class ConditionsRegister(dict):
+
```
Comment (brunobord) [inline formidable/forms/conditions.py]: missing docstring

File: formidable/forms/conditions.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    def build(self, fields, conditions_schema):
+        # TODO XXX deepcopy ?
+        return list(self.gen_from_schema(fields, conditions_schema))
+
+
+conditions_register = ConditionsRegister()
+
+
+class ConditionsMetaClass(type):
+
```
Comment (brunobord) [inline formidable/forms/conditions.py]: missing docstring

File: formidable/forms/conditions.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        klass = super(ConditionsMetaClass, mcls).__new__(
+            mcls, name, base, attrs
+        )
+        if 'action' in attrs:
+            conditions_register[klass.action] = klass
+        return klass
+
+
+class ConditionTest(object):
+
```
Comment (brunobord) [inline formidable/forms/conditions.py]: missing docstring

File: formidable/forms/conditions.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            ref_value = cleaned_data[self.field_id]
+            return meth(ref_value, self.values)
+        except:
+            # KeyError if self.field_id not in cleaned_data
+            # TODO XXX add log ?
+            return False
+
+
+class Condition(six.with_metaclass(ConditionsMetaClass)):
+
```
Comment (brunobord) [inline formidable/forms/conditions.py]: missing docstring

File: formidable/forms/conditions.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    def __init__(self, fields_ids, name, tests):
+        self.name = name
+        self.tests = tests
+        self.fields_ids = fields_ids
+
+
+class DisplayIffCondition(Condition):
+    action = 'display_iff'
```
Comment (brunobord) [inline formidable/forms/conditions.py]: missing docstring

File: demo/tests/test_conditions.py (hunk 2 above)
Comment (mgu) [inline demo/tests/test_conditions.py]: The payload generated by a "client" won't use our constant. The code checks the action is valid.
I'm not sure it's very useful here.
If still want a constant, I can add one.

File: demo/tests/test_conditions.py (hunk 3 above)
Comment (mgu) [inline demo/tests/test_conditions.py]: same answer

File: formidable/forms/conditions.py (hunk 6 above)
Comment (mgu) [inline formidable/forms/conditions.py]: Broken windows theory.
I started this using Presets as an example, but there's not a lot of docstring there...

File: formidable/serializers/forms.py [hunk 11]
```diff
@@ -2,25 +2,55 @@
...
 from django.db import transaction
+
 from formidable.models import Formidable
 from formidable.serializers import fields
 from formidable.serializers.common import WithNestedSerializer
 from formidable.serializers.presets import PresetModelSerializer
 from rest_framework import serializers
 from rest_framework.exceptions import ValidationError
 
+# TODO XXX pick choicefield values from formidable.conditions.*
```
Comment (brunobord) [inline formidable/serializers/forms.py]: I guess this TODO could be removed or transformed into a github issue

Comment (mgu) [inline formidable/serializers/forms.py]: fixed.

**Pull request merged**